library/bignum.c

```diff
@@ -2044,12 +2044,15 @@ static void mpi_montmul( mbedtls_mpi *A, const mbedtls_mpi *B, const mbedtls_mpi
     memcpy( A->p, d, ( n + 1 ) * ciL );
 
     /* If A >= N then A -= N. Do the subtraction unconditionally to prevent
-     * timing attacks. Modify T as a side effect. */
-    if( mbedtls_mpi_cmp_abs( A, N ) >= 0 )
-        mpi_sub_hlp( n, N->p, A->p );
-    else
-        /* prevent timing attacks */
-        mpi_sub_hlp( n, A->p, T->p );
+     * timing attacks. */
+    /* Set d to A + (2^biL)^n - N. */
+    d[n] += 1;
+    mpi_sub_hlp( n, N->p, d );
+    /* Now d - (2^biL)^n = A - N so d >= (2^biL)^n iff A >= N.
+     * So we want to copy the result of the subtraction iff d->p[n] != 0.
+     * Note that d->p[n] is either 0 or 1 since A - N <= N <= (2^biL)^n. */
+    mpi_safe_cond_assign( n + 1, A->p, d, d[n] );
+    A->p[n] = 0;
```

Comment:
I can see why this works and it is quite trivial compared to what came before, but maybe a clarifying comment would be helpful here as well.

Reply:
The most powerful reason why it works is that `A->p[n]` is already 0 here! (It's a leftover from a different calculation attempt that used A rather than d as a temporary.)

Reply:
Then I might be missing something, I thought that this would be either 1 or 0 here.

Reply:
No wait, am I confusing myself with my own code? if `d[n] == 1` then we're doing the assignment which sets `A->p[n]` to 1.

Reply:
Btw, typo in the comment: you're writing `d->p[n]` when you mean `d[n]` (twice).

Reply:
Let's recap calmly:

- If initially `A >= N`, then `d = 2^nbits + A - N >= 2^nbits`, ie `d[n] == 1` and the assignment is done, so `A_after = d = 2^nbits + A_before - N` has its top bit set which needs to be cleared so that `A_after = A_before - N`.
- If initially `A < N` then `d < 2^nbits`, ie `d[n] == 0`, the assignment is not done, `A_after` already has its top bit clear but it doesn't hurt to clear it again.

Reply:
I think it's probably easier to explain all at once rather than piecewise. The pieces of the computation don't actually make that much sense individually.

Reply:
I think the comments here need to be improved.

Reply:
I improved the comments and the code. In fact, there's never a need to touch `A->p[n]`: only the intermediate result `d` can go over n limbs.

Part of the confusion was that `mpi_sub_hlp` _looked_ like it only accessed n limbs of `d`, and I'd documented it as such, but that was in fact not the case. I've restructured the code to make `mpi_sub_hlp` only work inside the n limbs, and moved the carry propagation out into `mbedtls_mpi_sub_abs` and `mpi_montmul`. In `mpi_montmul`, the carry is directly subtracted from the next limb which we know is nonzero, which also avoids leaking the secret that we want to hide during carry propagation.